File: apps/blind_draft/server/ai.py
# -*- coding: utf-8 -*-
"""AI 对手页的数据装配 —— 只读。

**这一页不提供编辑。** 卡牌页那边人工层是主角(算法算错了要有出口),这边
恰恰相反:AI 的"当前实力"以前只能靠一条手拖的年龄曲线,`blinddraft/ai_teams.py`
自己写着「库里没有任何『这个人现在打得怎么样』的个人数据,所以只能设计,
不能查」——而现在 `5e_player_stats.json` 已有候选池近 12 个月的实测数据。
在有证据的地方开一个手调数值的口子,等于把刚拿到的证据又换回骰子。

所以这一页的职责是**把三列摆在一起**,给之后那套映射算法当工作台:

    卡面(生涯巅峰)   ← blinddraft/cards.py,玩家抽到的那张
    现况(AI 用的)    ← 当前角色 + 5E 当前火力 + 逐维生涯先验
    5E 实测          ← rating / adr / kast / kd / kpr / dpr / hs,近 12 个月

5E 当前证据现在只接管 Firepower；Leadership / Experience 继续走生涯证据，
Stability 在取得逐图方差前保留低置信先验。对应关系是逐维的，不是整卡替换。

队一级还摆了一个对照:我们算出来的 entry 顺位 vs HLTV 世界排名。
`blinddraft/major.py` 的注释里量过这两者的秩相关只有 0.53——那个数就是
这套映射要改善的东西,所以让它一直显示在页面上。
"""
import json

from playerdb.paths import BLIND_DRAFT, DATA

from blinddraft import ai_teams as A
from blinddraft import cards as C
from blinddraft import draft as P
from blinddraft import major as M

SNAP_PATH = BLIND_DRAFT / "team_snapshot.json"
STATS_PATH = BLIND_DRAFT / "5e_player_stats.json"
IMG5E_PATH = BLIND_DRAFT / "5e_images.json"
IMAGES_PATH = DATA / "images.json"

#: 5E 竞技数据里按数值用的字段。其余(first_blood / first_death / kddiff)是计数,
#: 单独走,不参与"越大越好"的着色。
STAT_NUM = ("rating", "adr", "kast", "kd", "kpr", "dpr", "win_rate")


def _load(path, default=None):
    if not path.exists():
        return default if default is not None else {}
    return json.loads(path.read_text(encoding="utf-8"))


def snapshot_index() -> dict:
    """5eplay 队伍快照,按队名和缩写都建一份索引(都折叠大小写)。"""
    out = {}
    for t in _load(SNAP_PATH).get("teams", []):
        for key in (t.get("name"), t.get("abbr")):
            if key:
                out.setdefault(key.casefold(), t)
    return out


def stats_index() -> tuple:
    """5E 个人数据。返回 (按 5e_id, 按 card_page) 两份索引。

    **主键是 5e_id**,因为当前世界里本来就有卡库没有的人——53 个候选池首发
    没有 card_page,按 page 索引就整批查不到。card_page 那份只是给还在按卡
    走的老路径用的桥。
    """
    by_id, by_page = {}, {}
    for sid, row in _load(STATS_PATH).get("players", {}).items():
        r = dict(row, _id=sid)
        by_id[sid] = r
        if row.get("card_page"):
            by_page[row["card_page"]] = r
    return by_id, by_page


def _num(v):
    try:
        return float(str(v).rstrip("%"))
    except (TypeError, ValueError):
        return None


def _photo(page, sid, img5e, lq):
    """优先 5E 的头像(当前世界的脸),没有就退回 Liquipedia 那套。"""
    if sid:
        p = img5e.get("players", {}).get(sid)
        if p:
            return "/bd/" + p
    p = lq.get("players", {}).get(page) if page else None
    return "/img/" + p if p else ""


def _spearman(pairs) -> float:
    """两个排名的秩相关。样本小(32 队),直接算,不引 scipy。"""
    n = len(pairs)
    if n < 3:
        return 0.0
    d2 = sum((a - b) ** 2 for a, b in pairs)
    return round(1 - 6 * d2 / (n * (n * n - 1)), 3)


def build_view(size=None) -> dict:
    """候选池 45 支队,阵容和位置直接来自队伍快照。

    以前这里是 `build_ai_field`:按 HLTV top100 挑队、从卡上的 `team` 字段拼
    阵容、位置靠猜、凑不满就补 G1 占位。三处都换掉了——**占位卡尤其**:
    The MongolZ 现在首发里三个人没打过 Major,补三张占位卡等于把这支队换成
    三个虚构的人。现在他们以本人身份进来,只是四维那一侧空着。

    卡库外真人现在也有 AI 专属的透明先验，所以每支五人完整的队都能计算 entry；
    他们的玩家卡一栏仍留白，不会因此进入玩家抽卡库。
    """
    cfg = M.load_config()
    snap_raw = _load(SNAP_PATH)
    cap = float(cfg.get("cohesion_cap", 4.0))
    idx = P.load_rosters()
    field, asof = A.build_pool_field(cfg)

    by_id, _by_page = stats_index()
    img5e = _load(IMG5E_PATH)
    lq = _load(IMAGES_PATH, {"players": {}, "flags": {}})
    career = {c["page"]: c for c in P.load_cards()}
    snap = snapshot_index()

    # 每个人现在都有 AI 当前四维；只有快照本身不满五人的队不能计算 entry。
    scored = []
    for t in field:
        if len(t["roster"]) == 5:
            r = A.entry_of(t, idx, cap)
            scored.append((r["entry"], t["name"], r))
    scored.sort(key=lambda x: -x[0])
    our = {name: i for i, (_e, name, _r) in enumerate(scored, 1)}
    rate = {name: r for _e, name, r in scored}
    # HLTV 顺位也只在同一批队里比：这个池子是按大区配额挑的,中间跳过的队
    # 会污染名次做差。
    hltv_order = {t["name"]: i for i, t in enumerate(
        sorted([t for t in field if t["name"] in our and t["rank"]],
               key=lambda t: t["rank"]), 1)}

    teams, ranks = [], []
    covered = total = nocard = 0
    for t in field:
        st = snap.get(t["name"].casefold()) or {}
        logo = img5e.get("teams", {}).get(t["id"], "") or img5e.get(
            "teams", {}).get(st.get("id", ""), "")
        roster = []
        for c in t["roster"]:
            sid = c.get("_5e_id")
            srow = by_id.get(sid or "")
            base = career.get(c.get("page") or "")
            total += 1
            if srow:
                covered += 1
            if c.get("_nocard"):
                nocard += 1
            roster.append({
                "page": c.get("page"), "nickname": c["nickname"],
                "position": c["position"], "grade": c.get("grade"),
                "caller": bool(c.get("caller", c["position"] == "IGL")),
                "age": c.get("age"), "country": c.get("country", ""),
                "nocard": bool(c.get("_nocard")),
                "role_src": c.get("_role_src"),
                "notes": c.get("_notes") or [],
                "cur": {k: c[k] for k in C.ATTRS} | {"overall": c["overall"],
                                                           "sources": c.get("_sources") or {}},
                "card": ({k: base[k] for k in C.ATTRS} | {"overall": base["overall"],
                          "position": base["position"], "grade": base["grade"]}
                         if base else None),
                "stat": ({k: _num(srow.get(k)) for k in STAT_NUM}
                         | {"hs_rate": _num(srow.get("hs_rate")),
                            "maps": _num(srow.get("map_count")),
                            "tier": srow.get("tier", ""),
                            "fb": _num(srow.get("first_blood")),
                            "fd": _num(srow.get("first_death"))}
                         if srow else None),
                "photo": _photo(c.get("page"), sid, img5e, lq),
                "flag": ("/img/" + lq["flags"][c["country"]]
                         if c.get("country") in lq.get("flags", {}) else ""),
            })
        pos = our.get(t["name"])
        r = rate.get(t["name"])
        ho = hltv_order.get(t["name"])
        evidence = {k: sum(1 for c in t["roster"]
                           if (c.get("_sources", {}).get("firepower", {})
                               .get("confidence")) == k)
                    for k in ("strong", "supporting", "low")}
        teams.append({
            "name": t["name"], "hltv": t["rank"], "vrs": t["vrs"],
            "region": t["region"], "seat": t["seat"], "stage": t["stage"],
            "our": pos, "hltv_order": ho,
            "delta": (ho - pos) if (ho and pos) else None,
            "entry": round(r["entry"], 1) if r else None,
            "evidence": evidence,
            "chem": round(r["chem_raw"], 1) if r else None,
            "cohesion": round(r["cohesion"], 1) if r else None,
            "adjust": t.get("adjust") or 0.0,
            "real": t["real"], "source": t["source"],
            "gaps": t.get("gaps"), "conflicts": t.get("conflicts"),
            "logo": "/bd/" + logo if logo else "",
            "roster": roster,
        })
        if ho and pos:
            ranks.append((ho, pos))

    slots = cfg.get("regional_slots") or {}
    return {
        "teams": teams,
        "skipped": [],
        "asof": asof,
        "cap": cap,
        "rho": _spearman(ranks),
        "scored": len(our),
        "coverage": {"with_stats": covered, "total": total, "nocard": nocard},
        "slots": {k: v for k, v in slots.items()},
        "pool": cfg.get("candidate_pool") or {},
        "snapshot_date": snap_raw.get("snapshot_date", ""),
        "partial_refresh_at": snap_raw.get("partial_refresh_at", ""),
        "partial_refresh_ids": snap_raw.get("partial_refresh_ids") or [],
        "age_curve": {"knee": A.AGE_KNEE, "rate": A.AGE_RATE, "exp": A.AGE_EXP},
        "stats_window": (_load(STATS_PATH).get("window") or {}).get("value", ""),
        "stats_grade": _load(STATS_PATH).get("grade_label", ""),
    }

File: apps/blind_draft/static/ai.js
/* AI 对手页 —— 只读。
 *
 * 一行一个人，三组数并排：
 *   卡面   生涯巅峰，玩家抽到的那张（blinddraft/cards.py）
 *   现况   AI 实际用的逐维当前卡；每格带来源和置信度（blinddraft/ai_teams.py）
 *   5E     近 12 个月的实测竞技数据（bdtools/fetch_5e_stats.py）
 *
 * 队伍是**候选池 45 支**：各大区 VRS 前 N，阵容和队内位置直接查队伍快照。
 * 卡库外真人的“卡面”留白，但“现况”可由当前证据 + 透明低置信先验生成；
 * 这不会把他们写进玩家卡库，也不再用匿名 G1 占位替换真人。
 */
'use strict';

const $ = (s) => document.querySelector(s);
const POS_CN = { RIFLER: '步枪', AWPER: '狙击', IGL: '指挥' };
const ATTRS = ['firepower', 'leadership', 'experience', 'stability'];

let DATA = null;
const open = new Set();

const esc = (s) => String(s == null ? '' : s).replace(/[&<>"]/g,
  (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]));
const n1 = (v) => (v == null ? '<span class="na">—</span>' : v.toFixed(1));
const n2 = (v) => (v == null ? '<span class="na">—</span>' : v.toFixed(2));

/* ------------------------------------------------------------------ 载入 */
async function load() {
  DATA = await (await fetch('/api/ai')).json();
  const c = DATA.coverage;
  const inField = DATA.teams.filter((t) => t.stage).length;
  $('#mRho').textContent = 'ρ ' + DATA.rho;
  $('#mCov').textContent = '5E 覆盖 ' + c.with_stats + '/' + c.total;
  $('#mAsof').textContent = '快照 ' + DATA.snapshot_date;

  const q = DATA.pool || {};
  const quota = ['欧洲', '美洲', '亚洲'].map((k) => {
    const v = (DATA.slots || {})[k];
    return v ? k[0] + ' ' + (v.stage1 + v.stage2 + v.stage3) : '';
  }).filter(Boolean).join(' / ');
  $('#banner').innerHTML = `
    <div>候选池 <b>${DATA.teams.length} 支</b>（欧洲 VRS 前 ${q['欧洲']} · 美洲前 ${q['美洲']} · 亚洲前 ${q['亚洲']}），
      其中 <b>${inField} 支</b>拿到 Major 名额（${esc(quota)}），
      其余 ${DATA.teams.length - inField} 支是留给 VRS 变动的余量。
      阵容与队内位置来自 <b>5eplay 快照（${esc(DATA.snapshot_date)}）</b>，
      实测窗口 <b>${esc((DATA.stats_window || '').replace('_', ' → '))}</b> · ${esc(DATA.stats_grade)}。
      默契上限 ${DATA.cap}。</div>
    <div class="warn"><b>${c.nocard} 人卡库里没有</b>——他们没有玩家卡，但有 AI 专属当前四维；
      当前证据不足的维度会明确标成低置信先验。五人完整的 <b>${DATA.scored} / ${DATA.teams.length}</b>
      支队均可计算预测种子分。</div>
    <div>当前 Firepower：80 图以上直接读标尺，30–79 图向生涯先验收缩；IGL 与无可靠证据者
      才使用年龄/生涯回退。预测顺位与 HLTV 排名秩相关 <b>ρ = ${DATA.rho}</b>，
      它是 Match 的赛前基准，不是官方实力排名。</div>`;
  if (location.hash === '#all') {
    $('#expandAll').checked = true;
    DATA.teams.forEach((t) => open.add(t.name));
  }
  render();
}

/* -------------------------------------------------------------- 队伍列表 */
function sorted() {
  const q = $('#q').value.trim().toLowerCase();
  const by = $('#sort').value;
  let out = DATA.teams.filter((t) => !q || t.name.toLowerCase().includes(q)
    || t.roster.some((p) => p.nickname.toLowerCase().includes(q)));
  const big = 1e6;                       // 缺值一律沉底，不参与比较
  const key = {
    vrs: (t) => t.vrs || big,
    our: (t) => (t.our == null ? big : t.our),
    hltv: (t) => t.hltv || big,
    delta: (t) => (t.delta == null ? big : -Math.abs(t.delta)),
  };
  return out.sort((a, b) => (key[by] || key.vrs)(a) - (key[by] || key.vrs)(b));
}

function teamHead(t) {
  const d = t.delta;
  const dTxt = d == null ? '<span class="na">—</span>'
    : d === 0 ? '<span class="dim">±0</span>'
    : `<span class="${d > 0 ? 'delta-up' : 'delta-dn'}">${d > 0 ? '+' : ''}${d}</span>`;
  const tags = [
    t.real < 5 ? `<span class="tag warn">${5 - t.real} 人卡库里没有</span>` : '',
    t.adjust ? `<span class="tag acc">人工 ${t.adjust > 0 ? '+' : ''}${t.adjust}</span>` : '',
    (t.gaps || []).length ? `<span class="tag warn">缺 ${t.gaps.join('/')}</span>` : '',
    t.evidence ? `<span class="tag">火力证据 ${t.evidence.strong}强/${t.evidence.supporting}弱/${t.evidence.low}先验</span>` : '',
  ].join('');
  // 区内名次才是决定名额的那个数，全球 VRS 只用来排种子。以前这里把全球名次
  // 标成「区内 VRS」，一直是错的。
  const sub = [t.region ? t.region + '区 #' + t.seat : '',
    t.stage ? 'Stage ' + t.stage : '余量',
    t.vrs ? '全球 VRS #' + t.vrs : ''].filter(Boolean).join(' · ');
  const num = (v, label) => `<div class="num"><span class="v">${v}</span><small>${label}</small></div>`;
  return `<div class="team-head${t.stage ? '' : ' reserve'}">
    <div class="pos">${t.our == null ? '<span class="na">·</span>' : t.our}</div>
    <div>${t.logo ? `<img class="logo" loading="lazy" src="${esc(t.logo)}">` : ''}</div>
    <div>
      <div class="nm">${esc(t.name)}${tags}</div>
      <div class="sub2">${esc(sub || '—')}</div>
    </div>
    ${num(t.entry == null ? '<span class="na">—</span>' : t.entry, '预测分')}
    ${num(t.cohesion == null ? '<span class="na">—</span>' : t.cohesion,
          t.chem != null && t.chem > t.cohesion ? `默契 · 原始 ${t.chem}` : '默契')}
    ${num(t.hltv ? '#' + t.hltv : '<span class="na">—</span>', 'HLTV')}
    ${num(dTxt, '顺位分歧')}
    <div class="lineup">${t.roster.map((p) => esc(p.nickname)).join(' · ')}</div>
  </div>`;
}

function playerRow(p) {
  const card = p.card, cur = p.cur, s = p.stat;
  // 现况按“比卡面高了还是低了”着色；悬停可看这一维的来源、置信度和解释。
  const cell = (k, extra) => {
    if (!cur) return `<td class="gB ${extra || ''} na">—</td>`;
    const cls = ['gB', extra || '', card && cur[k] < card[k] ? 'down'
                                  : card && cur[k] > card[k] ? 'up' : ''].join(' ').trim();
    const src = (cur.sources || {})[k] || {};
    const tip = [src.source, src.confidence, src.why].filter(Boolean).join(' · ');
    return `<td class="${cls}" title="${esc(tip)}">${cur[k]}</td>`;
  };
  const A = (v, extra) => `<td class="gA ${extra || ''}">${v}</td>`;
  const Cc = (v, extra) => `<td class="gC ${extra || ''}">${v}</td>`;
  return `<tr class="${p.nocard ? 'nocard' : ''}">
    <td class="l">${p.photo ? `<img class="face2" loading="lazy" src="${esc(p.photo)}">`
                            : '<div class="face2"></div>'}</td>
    <td class="l"><div class="pname">
      ${p.flag ? `<img class="flag" loading="lazy" src="${esc(p.flag)}">` : ''}
      <span>${esc(p.nickname)}</span></div></td>
    <td class="l dim">${POS_CN[p.position] || p.position}${
      p.caller && p.position !== 'IGL' ? '·指挥' : ''}${
      p.role_src ? `<i class="guess" title="位置由竞技数据判定：${esc(p.role_src)}">?</i>` : ''}</td>
    <td>${p.grade == null ? '<span class="na">—</span>'
                          : `<span class="g g${p.grade}">${p.grade}</span>`}</td>
    <td class="dim">${p.age == null ? '—' : p.age}</td>

    ${card ? A(card.firepower, 'sep') + A(card.leadership) + A(card.experience)
             + A(card.stability) + A(card.overall)
           : A('—', 'sep na') + A('—', 'na') + A('—', 'na') + A('—', 'na') + A('—', 'na')}

    ${cell('firepower', 'sep')}${cell('leadership')}${cell('experience')}${cell('stability')}
    <td class="gB">${cur ? `<b>${Number(cur.overall).toFixed(1)}</b>`
                         : '<span class="na">—</span>'}</td>

    ${s ? Cc(n2(s.rating), 'sep') + Cc(n1(s.adr)) + Cc(n1(s.kast)) + Cc(n2(s.kd))
          + Cc(n2(s.kpr)) + Cc(n2(s.dpr)) + Cc(n1(s.hs_rate))
          + Cc(s.maps == null ? '—' : s.maps, 'dim')
          + Cc(esc(s.tier), 'dim')
        : '<td class="gC sep na" colspan="9">这一年没有 Major / S+ / S 级样本</td>'}

    <td class="l note2">${p.notes.map(esc).join('；')}</td>
  </tr>`;
}

function rosterTable(t) {
  return `<div class="roster"><table class="ros">
    <thead>
      <tr>
        <th colspan="5"></th>
        <th class="grp gA sep" colspan="5">卡面 · 生涯巅峰</th>
        <th class="grp gB sep" colspan="5">现况 · AI 实际用的</th>
        <th class="grp gC sep" colspan="9">5E 实测 · 近 12 个月</th>
        <th></th>
      </tr>
      <tr>
        <th></th><th class="l">选手</th><th class="l">位</th><th>档</th><th>岁</th>
        <th class="gA sep">火</th><th class="gA">领</th><th class="gA">经</th>
        <th class="gA">稳</th><th class="gA">OVR</th>
        <th class="gB sep">火</th><th class="gB">领</th><th class="gB">经</th>
        <th class="gB">稳</th><th class="gB">OVR</th>
        <th class="gC sep">rating</th><th class="gC">ADR</th><th class="gC">KAST</th>
        <th class="gC">K/D</th><th class="gC">KPR</th><th class="gC">DPR</th>
        <th class="gC">HS%</th><th class="gC">图数</th><th class="gC">档次</th>
        <th class="l">改动</th>
      </tr>
    </thead>
    <tbody>${t.roster.map(playerRow).join('')}</tbody>
  </table></div>`;
}

function render() {
  const list = sorted();
  $('#count').textContent = list.length + ' 支队';
  $('#teams').innerHTML = list.map((t) => `
    <div class="team ${open.has(t.name) ? 'open' : ''}" data-name="${esc(t.name)}">
      ${teamHead(t)}
      ${open.has(t.name) ? rosterTable(t) : ''}
    </div>`).join('');

  $('#skipped').innerHTML = (DATA.skipped || []).length ? `
    <h4>没进赛场的队（现役队员不足，需要在 major_field.json 里手写阵容）</h4>
    ${DATA.skipped.map((s) =>
      `<span>HLTV #${s.rank} <b>${esc(s.name)}</b> 只有 ${s.have} 人</span>`).join('')}` : '';
}

/* -------------------------------------------------------------------- 绑定 */
$('#q').addEventListener('input', render);
$('#sort').addEventListener('change', render);
$('#expandAll').addEventListener('change', (e) => {
  open.clear();
  if (e.target.checked) DATA.teams.forEach((t) => open.add(t.name));
  render();
});
$('#teams').addEventListener('click', (e) => {
  const box = e.target.closest('.team');
  if (!box) return;
  const name = box.dataset.name;
  open.has(name) ? open.delete(name) : open.add(name);
  render();
});

load().catch((e) => { $('#banner').textContent = '载入失败：' + e.message; });
